src/WebJobs.Script/Binding/Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -0,0 +1,320 @@
+        }
+
+        internal class HttpTriggerBinding : ITriggerBinding
+        {
+            private readonly ParameterInfo _parameter;
+            private readonly IBindingDataProvider _bindingDataProvider;
+            private readonly bool _isUserTypeBinding;
+
+            public HttpTriggerBinding(ParameterInfo parameter, bool isUserTypeBinding)
+            {
+                _parameter = parameter;
+                _isUserTypeBinding = isUserTypeBinding;
+
+                if (_isUserTypeBinding)
+                {
+                    // Create the BindingDataProvider from the user Type. The BindingDataProvider
+                    // is used to define the binding parameters that the binding exposes to other
+                    // bindings (i.e. the properties of the POCO can be bound to by other bindings).
+                    // It is also used to extract the binding data from an instance of the Type.
+                    _bindingDataProvider = BindingDataProvider.FromType(parameter.ParameterType);
+                }
+            }
+
+            public IReadOnlyDictionary<string, Type> BindingDataContract
+            {
+                get
+                {
+                    // if we're binding to a user Type, we'll have a contract,
+                    // otherwise none
+                    return _bindingDataProvider != null ? _bindingDataProvider.Contract : null;
+                }
+            }
+
+            public Type TriggerValueType
+            {
+                get { return typeof(HttpRequestMessage); }
+            }
+
+            public async Task<ITriggerData> BindAsync(object value, ValueBindingContext context)
+            {
+                HttpRequestMessage request = value as HttpRequestMessage;
+                if (request == null)
+                {
+                    throw new NotSupportedException("An HttpRequestMessage is required");
+                }
+
+                IValueProvider valueProvider = null;
+                IReadOnlyDictionary<string, object> bindingData = null;
+                string invokeString = request.ToString();
+                if (_isUserTypeBinding)
+                {
+                    valueProvider = await CreateUserTypeValueProvider(request, invokeString);
+                    if (_bindingDataProvider != null)
+                    {
+                        // binding data is defined by the user type
+                        // the provider might be null if the Type is invalid, or if the Type
+                        // has no public properties to bind to
+                        bindingData = _bindingDataProvider.GetBindingData(valueProvider.GetValue());
+                    }
+                }
+                else
+                {
+                    valueProvider = new HttpRequestValueBinder(_parameter, request, invokeString);
+                    bindingData = GetRequestBindingData(request);
```

Comment:
When NOT binding to a POCO, you can see here I'm still extracting some binding data from the request. That is for the non C# scenarios. In C# and other early bound cases, functions that attempt to use binding expressions w/o a POCO trigger type will fail indexing (no static binding contract).